polkadot/consensus/src/lib.rs

```diff
@@ -679,45 +514,181 @@ impl<C: PolkadotApi, R: TableRouter> bft::Proposer for Proposer<C, R> {
+		if included == self.last_included {
+			return self.enough_candidates.poll();
+		}
+
+		// the amount of includable candidates has changed. schedule a wakeup
+		// if it's not sufficient anymore.
+		match self.dynamic_inclusion.acceptable_in(Instant::now(), included) {
+			Some(duration) => {
+				self.last_included = included;
+				self.enough_candidates = self.timer.sleep(duration);
+				self.enough_candidates.poll()
+			}
+			None => {
+				Ok(Async::Ready(()))
+			}
+		}
 	}
+}
+
+/// Future which resolves upon the creation of a proposal.
+pub struct CreateProposal<C: PolkadotApi, R, P: Collators>  {
+	parent_hash: HeaderHash,
+	parent_number: BlockNumber,
+	parent_id: C::CheckedBlockId,
+	client: Arc<C>,
+	transaction_pool: Arc<Mutex<TransactionPool>>,
+	collation: CollationFetch<P, C>,
+	router: R,
+	table: Arc<SharedTable>,
+	timing: ProposalTiming,
+}
+
+impl<C, R, P> CreateProposal<C, R, P>
+	where
+		C: PolkadotApi,
+		R: TableRouter,
+		P: Collators,
+{
+	fn propose_with(&self, candidates: Vec<CandidateReceipt>) -> Result<SubstrateBlock, Error> {
+		// TODO: handle case when current timestamp behind that in state.
+		let timestamp = current_timestamp();
+		let mut block_builder = self.client.build_block(
+			&self.parent_id,
+			timestamp,
+			candidates,
+		)?;
+
+		let readiness_evaluator = Ready::create(self.parent_id.clone(), &*self.client);
+
+		{
+			let mut pool = self.transaction_pool.lock();
+			let mut unqueue_invalid = Vec::new();
+			let mut pending_size = 0;
+			for pending in pool.pending(readiness_evaluator.clone()) {
+				// skip and cull transactions which are too large.
+				if pending.encoded_size() > MAX_TRANSACTIONS_SIZE {
+					unqueue_invalid.push(pending.hash().clone());
+					continue
+				}
+
+				if pending_size + pending.encoded_size() >= MAX_TRANSACTIONS_SIZE { break }
+
+				match block_builder.push_extrinsic(pending.as_transaction().clone()) {
+					Ok(()) => {
+						pending_size += pending.encoded_size();
+					}
+					Err(_) => {
+						unqueue_invalid.push(pending.hash().clone());
+					}
+				}
+			}
+
+			for tx_hash in unqueue_invalid {
+				pool.remove(&tx_hash, false);
+			}
+		}
+
+		let polkadot_block = block_builder.bake();
 
-	let block_timestamp = proposal.timestamp();
+		info!("Proposing block [number: {}; hash: {}; parent_hash: {}; extrinsics: [{}]]",
+			polkadot_block.header.number,
+			Hash::from(polkadot_block.header.blake2_256()),
+			polkadot_block.header.parent_hash,
+			polkadot_block.extrinsics.iter()
+				.map(|xt| format!("{}", Hash::from(xt.blake2_256())))
+				.collect::<Vec<_>>()
+				.join(", ")
+		);
 
-	// TODO: just defer using `tokio_timer` to delay prepare vote.
-	if block_timestamp > now + MAX_TIMESTAMP_DRIFT {
-		bail!(ErrorKind::TimestampInFuture)
+		let substrate_block = Slicable::decode(&mut polkadot_block.encode().as_slice())
+			.expect("polkadot blocks defined to serialize to substrate blocks correctly; qed");
+
+		// TODO: full re-evaluation
```

Comment:
not sure how necessary full re-evaluation is here given that the BFT code currently has proposers re-evaluate their own blocks anyway.